### PR TITLE
Quotas

### PR DESCRIPTION
**Pull request by byucesoy**
**Add helper methods for quota calculations**
I was not entirely sure about whether to put default quotas as config file or
as a database table. I decided to go with a config file for now, because
- It is easier to change. Does not require a migration.
- It is easier to view. You can see all the quotas in one place.
- It seldom changes. Control plane does not need to edit default quotas as
part of its normal operation.

The drawback is that, we cannot join with it, so I had to write a helper
function to calculate effective_quota_value. Though, it is small enough that
I'm not sure if the join version would be any better.

**Add quota check for VM provisionings**
There are two important notes about this commit.
1. We don't differentiate between x64 and arm64 VMs for now, as it is not
possible to request arm64 VMs yet. In all cases, we calculate the core count as
vCPU / 2. The logic would need to change when we introduce arm64 VMs.

2. There is one weakness that I'd like to address, but it would touch APIs and
would require a deeper discussion, so I'm leaving that for future. In summary
when the vm size is not specified by user, we default to standard-2, however,
the resolution is done at the prog's assemble method. We are doing the quota
check at the API layer, thus we don't know the size of the VM at this point
(if it is not specified by user). I added a branch to calculate requested cores
as 1 if the size is not specified, but this means we leak information from prog
to API. I'd like to avoid this and I think cleanest solution would be making
VM size a required field. Apart from fixing this problem, I think it would make
more sense semantically as well. It feels bit weird that Vm size is optional
in Vm creation API.

**Add quota check for Github runner provisionings**
We are calling quota_available? method with 0 as requested_additional_usage.

`project.quota_available?("GithubRunnerCores", 0)`

This looks weird, but it is intentional. In existing Github quota calculations,
we compare the total allocated cores with the core limit and allow passing the
limit once. This is because we check quota and allocate VMs in different labels
(hence different transactions) and it is difficult to enforce the quotas in the
environment with lots of concurrent requests. There are some remedies, but it
would require some refactoring that I'm not keen to do at the moment. Passing 0
as requested_additional_usage keeps the existing behavior.

**Add quota check for Postgres provisionings**
Nothing fancy, we just check the current usage and allow provisioning if the
request would not cause quota to be exceeded.

### CODE REVIEW

File: spec/model/project_spec.rb [hunk 1]
```diff
@@ -74,4 +74,32 @@
       expect(project.default_location).to eq Option.locations.first.name
     end
   end
+
+  it "calculates current resource usage" do
```
Comment (enescakir) [inline spec/model/project_spec.rb:78]: Could you add a test to confirm that all quota IDs are unique, similar to what we do in billing rates?

File: config/default_quotas.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+- { id: c088f732-0a30-454b-aa6b-542ce1d67bfb, resource_type: VmCores,           value: 8   }
+- { id: 14fa6820-bf63-41d2-b35e-4a4dcefd1b15, resource_type: GithubRunnerCores, value: 150 }
```
Comment (enescakir) [inline config/default_quotas.yml:2]: Can we assert that all `id` and `resource_type` values are unique? Alternatively, could there be multiple instances of the same `resource_type` with different IDs? Is there any scenario for that?

File: model/project.rb [hunk 3]
```diff
@@ -98,6 +99,26 @@ def current_invoice
...
+    when "GithubRunnerCores" then github_installations.sum(&:total_active_runner_cores)
+    when "PostgresCores" then postgres_resources.flat_map { _1.private_subnet.vms.map(&:cores) }.sum
+    else
+      raise "Unknown resource type: #{resource_type}"
+    end
+  end
+
+  def effective_quota_value(resource_type)
+    default_quota = ProjectQuota.default_quotas[resource_type]
+    overide_quota_value = quotas_dataset.where(quota_id: default_quota["id"]).first&.value
```
Comment (enescakir) [inline model/project.rb:114]: nit
```suggestion
    overide_quota_value = quotas_dataset.first(quota_id: default_quota["id"])&.value
```

Comment (enescakir) [inline model/project.rb:114]: Is it a typo? I forgot to correct one 😄 
```suggestion
    override_quota_value = quotas_dataset.where(quota_id: default_quota["id"]).first&.value
```

File: model/project.rb [hunk 4]
```diff
@@ -98,6 +99,26 @@ def current_invoice
     Invoice.new(project_id: id, content: content, begin_time: begin_time, end_time: end_time, created_at: Time.now, status: "current")
   end
 
+  def current_resource_usage(resource_type)
+    case resource_type
+    when "VmCores" then vms.sum(&:cores)
+    when "GithubRunnerCores" then github_installations.sum(&:total_active_runner_cores)
+    when "PostgresCores" then postgres_resources.flat_map { _1.private_subnet.vms.map(&:cores) }.sum
```
Comment (enescakir) [inline model/project.rb:106]: Can't we use `PostgresServer` relation? I find it more intuitive.
```suggestion
    when "PostgresCores" then postgres_resources.flat_map { _1.servers.sum { |ps| ps.vm.cores } }.sum
```

File: prog/vm/github_runner.rb [hunk 5]
```diff
@@ -165,9 +165,16 @@ def before_run
...
     github_runner.installation.project_dataset.for_update.all
-    github_runner.installation.project.runner_core_limit > github_runner.installation.project.github_installations.sum(&:total_active_runner_cores)
+    # In existing Github quota calculations, we compare total allocated cores
+    # with the core limit and allow passing the limit once. This is because we
+    # check quota and allocate VMs in different labels hence transactions and it
+    # is difficult to enforce quotas ih the environment with lots of concurrent
+    # requests. There are some remedies, but it would require some refactoring
+    # that I'm not keen to do at the moment. Although it looks weird, passing 0
+    # as requested_additional_usage keeps the existing behavior.
+    github_runner.installation.project.quota_available?("GithubRunnerCores", 0)
```
Comment (enescakir) [inline prog/vm/github_runner.rb:177]: If I merge my pull request before yours. https://github.com/ubicloud/ubicloud/pull/1782

```suggestion
    project.quota_available?("GithubRunnerCores", 0)
```

File: prog/vm/github_runner.rb [hunk 6]
```diff
@@ -165,9 +165,16 @@ def before_run
...
   end
 
-  def concurrency_available?
+  def quota_available?
     github_runner.installation.project_dataset.for_update.all
-    github_runner.installation.project.runner_core_limit > github_runner.installation.project.github_installations.sum(&:total_active_runner_cores)
+    # In existing Github quota calculations, we compare total allocated cores
+    # with the core limit and allow passing the limit once. This is because we
+    # check quota and allocate VMs in different labels hence transactions and it
+    # is difficult to enforce quotas ih the environment with lots of concurrent
```
Comment (enescakir) [inline prog/vm/github_runner.rb:173]: Typo `the quotas ih the environment`

File: prog/vm/github_runner.rb [hunk 7]
```diff
@@ -165,9 +165,16 @@ def before_run
...
+    # requests. There are some remedies, but it would require some refactoring
```
Comment (enescakir) [inline prog/vm/github_runner.rb:174]: > it is difficult to enforce quotas ih the environment with lots of concurrent requests.

We have already locked the project using `for_update`. Isn't it enforced?

File: routes/common/postgres_helper.rb [hunk 8]
```diff
@@ -34,6 +34,9 @@ def post(name: nil)
     request_body_params = Validation.validate_request_body(params, required_parameters, allowed_optional_parameters)
     parsed_size = Validation.validate_postgres_size(request_body_params["size"])
 
+    requested_postgres_core_count = parsed_size.vcpu / 2
+    Validation.validate_core_quota(project, "PostgresCores", requested_postgres_core_count)
```
Comment (enescakir) [inline routes/common/postgres_helper.rb:38]: If HA is enabled, should we also consider the standby core counts?

File: migrate/20240720_project_quota.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+Sequel.migration do
+  up do
+    create_table(:project_quota) do
+      column :project_id, :uuid, null: false
+      column :quota_id, :uuid, null: false
+      column :value, :integer, null: false
+      primary_key [:project_id, :quota_id]
+    end
+
+    run "INSERT INTO project_quota (SELECT id, '14fa6820-bf63-41d2-b35e-4a4dcefd1b15', runner_core_limit AS quota_id FROM project WHERE runner_core_limit != 150)"
```
[the remark this answers is not in the document]
Comment (byucesoy) [reply migrate/20240720_project_quota.rb:12]: It was typo, removed it.

File: config/default_quotas.yml (hunk 2 above)
Comment (byucesoy) [reply config/default_quotas.yml:2]: In the future, we might have other fields to split resource types. For now they are unique, so I added a test case.

File: prog/vm/github_runner.rb (hunk 7 above)
Comment (byucesoy) [reply prog/vm/github_runner.rb:174]: Yes, but after we process the label, we commit the transaction, which releases the lock. We pick the VM in another state. If quota is available, multiple progs can move to `allocate_vm`

File: prog/vm/github_runner.rb [hunk 10]
```diff
@@ -125,12 +125,12 @@ def before_run
   end
 
   label def start
-    hop_wait_concurrency_limit unless concurrency_available?
+    hop_wait_concurrency_limit unless quota_available?
```
Comment (fdr) [inline prog/vm/github_runner.rb:128]: " Add quota check for Github runner provisionings" it seems like the commit previous to this one (that adds "dark" code uncalled), pause, and then deploying "Add quota check" would be a somewhat safe deploy (& rollback) strategy.

**Review by fdr — APPROVED** [review]
I thought it was informatively justified (reading patch by patch) and phased all around.

**Review by enescakir — APPROVED** [review]

**Pull request merged**